### DISCUSSION TITLE
Why do ContractOffers require a "1:" at the start of their ID?

### DISCUSSION THREAD

**Discussion by mark-olchanski** [eclipse-edc/Connector#937] [Q&A] [answered]
As the title says, I can only bring the Contract Negotiation process to work when the Contract Offer-ID starts with a "1:". What is the intention behind this?

**Comment by DominikPinsel** [comment, marked as answer]
The ID of the contract offer consists of a _contract definition id_ and a _unique offer part_, that are sparated by a semicolon. E.g. `[contractDefinitionId]:[uuid]`.
The reason you are finding a '1' there is, that the sample `ContractDefinition`, that was used to generate the offer, has probably the id '1'.

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

**Reply by ndr-brt** [reply]
Yes. I was wondering what was the reason for this choice, in my opinion would be better to separate those id in two different fields, giving to the agreement its own id.

**Reply by DominikPinsel** [reply]
The `ContractDefinition` id is there is to find the corresponding `ContractDefinition` in case of a future negotation for a `ContractOffer`(which are not persisted as long as no negotiation exist). The second part is there to have a unique ID, as it would be possible to generate thousands of `ContractOffers` from a single `ContractDefinition`.
But you are probably right that there might be a better solution. Putting the `ContractDefinition` in the offer-id feels like a workaround.

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

**Reply by mark-olchanski** [reply]
Thank you for your explanation. But how come that I can't perform a successful contract negotiation with any other ```ContractDefinition```? Doing something along the line of ```id("${contractDefinition.id}:${UUID.randomUUID()}")``` still resultes in a rejection.

**Reply by DominikPinsel** [reply]
At the moment all `ContractOffers` that do not match **excatly** an already existing `ContractOffer` are rejected. Therefore, initiating a custom negotiation, that does not contain something that is already offered, is not possible.

**Reply by jimmarino** [reply]
> As the title says, I can only bring the Contract Negotiation process to work when the Contract Offer-ID starts with a "1:". What is the intention behind this?

Contract offers by design must originate from the provider, particularly given that allowing consumer-generated contract offers for data originating at a provider is a security risk. The contract offer id is opaque from the standpoint of the client. The fact that it is a concatenated id is an implementation detail of the EDC and not relevant to the client.

For contract negotiation to work, the client must pass back an id that was provided in a previous request.
